### PR TITLE
remote-week2: Node.JS APIs

### PR DESCRIPTION
**Pull request by raamiiChu**
# AWS  
- Elastic IP Address: `52.192.99.138`  

- Database Name: `assignment`  

- Health Check API: http://52.192.99.138:3000/healthcheck  

---  

# Change Summary  
1. Add `remote-week2` folder  
    - `app.js`: Connect to the database with Node.JS    
    - `backup.sql`: Database backup file  

2. Add `appspec.yml` & `scripts` folder  
  For code deployment on AWS

---  

# Understanding Check  

### How did you design the user table schema? Why?  

```
create table `users`(
	id int PRIMARY KEY AUTO_INCREMENT,
	name varchar(32) NOT NULL,
	email varchar(64) NOT NULL UNIQUE,
	password varchar(64) NOT NULL,
	created_at datetime DEFAULT now()
);
```

- id: `PRIMARY KEY`, each data should have one unique primary key to recognize it. Set it as `AUTO_INCREMENT` can make sure the value can auto-generated.   

- name: Can not be empty  

- email: Each email can only register one account, so I set it as `UNIQUE`  

- password: Can not be empty  

- created_at: Set its default value as current time  

---  

### What does the *Status Code* mean? Is it a must to use?  

**Status Code** is the number which is returned by the server to response the request of client. **Status Code** has 5 kinds, each represented by the first digit of the code:  

- 1XX: Informational  
- 2XX: Successful  
- 3XX: Redirection  
- 4XX: Client Error  
- 5XX: Server Error  

**State Code** provide a standard way to tranfer the result of a request, making it easier for developers to analyze and handle issues. Therefore, it's a fundamental part of the HTTP / HTTPS protocol. Must to use it when developing a web server or an API.  

---  

### How and what tools did you use to test your APIs?  

[**VS Code Thunder Client**](https://marketplace.visualstudio.com/items?itemName=rangav.vscode-thunder-client)  
It's an extension in VS Code providing a easy way for developers to test their APIs without switching to other apps.    

VS Code Thunder Client

---  

### How and what tools did you use to make sure the app would not shut down even if you exit the ssh connection?  

[**Details**](https://www.notion.so/Code-Deployment-91a5ec6e84744ddaa1773001523586e6?pvs=4)

I use CodeDelpoy & CodePipeline in AWS to deploy the code automatically, the method can make sure the app would not shut down even if exiting the ssh connection. As for `.env` file, I choose to upload it manually, due to its sensitive info.  

---  

### What are the interactions between client, firewall, EC2 and RDS when you send a request to you app server and get the response? You can use sequence diagram to describe it more clearly.  

AWS Sequence Diagram
Image created with [Canva](https://www.canva.com/)

---

### ( Advanced Optional ) What did you do to make clients access your APIs without the certain port number?

Just Change the port number to 80 / 443, beacuse it's the default port number of HTTP / HTTPS. 

p.s. I do not push the code to GitHub

```
const app = express();
const port = 80;
```

**Connect without port number**
Connect without port number

---  

### ( Advanced Optional ) Compare to the application server, what's the web server?

- Web Server
    Handle HTTP / HTTPS requests and responses. It serves static content, such as HTML, CSS, images, etc.

- Application Server
    Execute dynamic content and application functionality. It handles dynamic tasks, such as database interactions, user authentication, etc.

---

### CODE REVIEW

File: remote-week2/app.js [hunk 1]
```diff
@@ -0,0 +1,177 @@
+require("dotenv").config();
+
+const express = require("express");
+const bodyParser = require("body-parser");
```
Comment (addie-tyc) [inline remote-week2/app.js]: Express 4.18 以後還有需要引入 body-parser 嗎？

File: remote-week2/app.js [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+// middleware
+app.use(bodyParser.json());
+app.use((req, res, next) => {
+    req.headers["content-type"] = "application/json";
+
+    req.headers["request-date"] = new Date().toISOString();
+
+    next();
+});
```
Comment (addie-tyc) [inline remote-week2/app.js]: 其實文件中 Request-Date 會由 client 帶進去，用 middleware 和題意稍有不同，
content-type 的部分通常會由 client 指定，後端 server 會負責驗證這個請求是否需要被處理

File: remote-week2/app.js [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+// Create the connection to database
+async function connectToMysql(res) {
+    try {
+        const connection = await mysql.createConnection({
+            host: process.env.APPWORK_RDS_HOST,
+            port: process.env.APPWORK_RDS_PORT,
+            user: process.env.APPWORK_RDS_USER,
+            database: process.env.APPWORK_RDS_DATABASE,
+            password: process.env.APPWORK_RDS_PASSWORD,
+        });
```
Comment (addie-tyc) [inline remote-week2/app.js]: 用 .env 滿好的！

**Review by addie-tyc — CHANGES_REQUESTED** [review]
code 的部分沒有問題，但剛剛檢測 API 到一半突然連 healthcheck 都沒打過？
部署的部分其實我沒有用過 codePipeline，要注意免費用量喔！

File: remote-week2/app.js (hunk 1 above)
Comment (raamiiChu) [reply remote-week2/app.js]: 新版直接用 express 內建的函式即可

```
app.use(express.json());
```

File: remote-week2/app.js (hunk 2 above)
Comment (raamiiChu) [reply remote-week2/app.js]: 程式碼保留（經過測試，註解程式碼後若格式撰寫正確，能完成題目原先要求）

- Content-Type
Content-Type

- Request-Date
    ```
    var currentDate = new Date().toUTCString();
    pm.request.headers.add({ key: 'request-date', value: currentDate });
    ```
    request-date

**Review by addie-tyc — APPROVED** [review]
LGTM. 有 hash 不錯

**Review by addie-tyc — APPROVED** [review]
LGTM

**Pull request merged**